### PR TITLE
AB#4218 - Make device type more specific

### PR DESCRIPTION
**Pull request by mknoedel**
### Changed

- The type of `intune_managed_device` was changed to more accurately describe
  the hardware of the device. The new device types are `user_endpoint`,
  `workstation`, `laptop`, `desktop`, `computer`, `smartphone`, `tablet`
- The class of `intune_managed_device` only includes `Device` if the device is
  not a virtual machine

### Added

- `deviceType` and `function` keys to `intune_managed_device` entities

### CODE REVIEW

File: docs/jupiterone.md [hunk 1]
```diff
@@ -136,16 +142,52 @@ The following relationships are created/mapped:
...
+| `computer`                    | **ASSIGNED**          | `intune_compliance_policy`     |
+| `computer`                    | **ASSIGNED**          | `intune_managed_application`   |
+| `computer`                    | **HAS**               | `intune_detected_application`  |
+| `computer`                    | **HAS**               | `intune_noncompliance_finding` |
+| `computer`                    | **USES**              | `intune_device_configuration`  |
+| `desktop`                     | **ASSIGNED**          | `intune_compliance_policy`     |
+| `desktop`                     | **ASSIGNED**          | `intune_managed_application`   |
+| `desktop`                     | **HAS**               | `intune_detected_application`  |
+| `desktop`                     | **HAS**               | `intune_noncompliance_finding` |
+| `desktop`                     | **USES**              | `intune_device_configuration`  |
```
Comment (ndowmon) [inline docs/jupiterone.md:161]: Might be worth doing something similar to what we do in the `graph-azure` project, where we set up the docs to say `ANY_SCOPE -HAS-> azure_diagnostic_log_setting` (and a few other resources): https://github.com/JupiterOne/graph-azure/blame/master/docs/jupiterone.md#L226

Maybe a pattern worth considering here.

File: src/steps/intune/constants.ts [hunk 2]
```diff
@@ -45,27 +65,47 @@ export const entities: { [k: string]: StepEntityMetadata } = {
     _type: 'intune_detected_application',
     _class: 'Application',
   },
-};
+} as const;
```
Comment (ndowmon) [inline src/steps/intune/constants.ts:68]: Seems random... necessary?

File: src/steps/intune/steps/devices/converters.ts [hunk 3]
```diff
@@ -78,13 +88,61 @@ export function createManagedDeviceEntity(
...
+    case 'palm':
+      return 'smartphone';
+    case 'winCE':
+    case 'winEmbedded':
+      return 'workstation';
+    case 'macMDM':
+    case 'windows10x':
+      return 'laptop';
+    case 'unix': // don't know where to put this but seems like a good guess
+    case 'linux': // don't know where to put this but seems like a good guess
```
Comment (ndowmon) [inline src/steps/intune/steps/devices/converters.ts:132]: weird that these are operating systems (or specs for OSs). Any other properties that could help us determine if this is physical or not?

**Review by ndowmon — DISMISSED** [review]

File: src/steps/intune/constants.ts (hunk 2 above)
Comment (mknoedel) [reply src/steps/intune/constants.ts:68]: Yeah, it's needed for typescript to not be annoying when dealing with some values being arrays and some being strings. `as const` makes it so typescript knows the exact type of a key which saves the `as string` and `as string[]` checks

File: src/steps/intune/steps/devices/converters.ts (hunk 3 above)
Comment (mknoedel) [reply src/steps/intune/steps/devices/converters.ts:132]: Hmm, thinking about it again, I can tell if they are physical or not, but I can't be certain about if they are a desktop or a laptop... I think what I am going to do is check if they are physical, then they will be `desktop`s and if not they are `server`s

Comment (mknoedel) [reply src/steps/intune/steps/devices/converters.ts:132]: - [ ]  Type `unix` and `linux` devices as `desktop` and `server`

File: docs/jupiterone.md (hunk 1 above)
Comment (mknoedel) [reply docs/jupiterone.md:161]: Interesting, though I think in this case it likely is fine as is, as large amount of these relationships get moved to the `intune-host-agent` in the next PR. Definitely something to keep in mind though!

File: src/steps/intune/steps/applications/index.ts [hunk 4]
```diff
@@ -66,30 +72,37 @@ export async function fetchDetectedApplications(
...
-              }),
-            );
-          }
+                await jobState.addRelationship(
+                  createDirectRelationship({
+                    _class:
+                      relationships.MULTI_DEVICE_HAS_DETECTED_APPLICATION[0]
+                        ._class,
+                    from: deviceEntity,
+                    to: detectedAppEntity,
```
Comment (aiwilliams) [inline src/steps/intune/steps/applications/index.ts:97]: I think we'll need to make another PR that creates a mapped relationship instead of creating the Application and a direct relationship, so the result is:

1. A single `Application` entity for each application name, with no version property (or other device specific properties) on the application
2. The device -> application relationship has the version installed on that device

**Review by aiwilliams — APPROVED** [review]

File: src/steps/intune/steps/applications/index.ts (hunk 4 above)
Comment (mknoedel) [reply src/steps/intune/steps/applications/index.ts:97]: Good point, that makes a ton of sense. Story written.

File: src/steps/intune/constants.ts (hunk 2 above)
Comment (mknoedel) [reply src/steps/intune/constants.ts:68]: I believe the JAMF solution wouldn't fix the weird typing issues when each of these keys could be either `StepEntityMetadata` or `StepEntityMetadata[]`. Const fixes that problem while still maintaining type support.
